Conversation Engine Refactor

kauffecup opened the pull request:
> Not finished, not running, up next is the "get next conversation block" logic. I also need to change item requirements from being exit requirements for a text block to entrance requirements (change from "you need x to exit this part of the conversation to you need x to enter the next part of the conversation) since we may want only some branches to have item requirements

Commit:
Not finished, not running, up next is the "get next conversation block" logic. I also need to change item requirements from being exit requirements for a text block to entrance requirements (change from "you need x to exit this part of the conversation to you need x to enter the next part of the conversation) since we may want only some branches to have item requirements

kauffecup commented:
![bitmoji](https://render.bitstrips.com/v2/cpanel/9853640-128732286_4-s4-v1.png?palette=1&width=246)

Commit:
Totally redoing the advance conversation block in Area, still need to move item requirements to the blocks players are trying to enter. Trying to add new functions in Dialogue instead of changing existing ones, will do a pass at the end and remove unused functions

Commit:
Working (as far as I can tell) fancy dialogue. Text files still need to be updated and we need to test more to make sure I didn't break battles or something non-obvious

jordantoth commented:
Ready for some testing and review

kauffecup commented on src/Dialogue.js:
```diff
@@ -15,6 +15,15 @@ export default class Dialogue {
+    for (var text in this.conversation) {
+      this.activeText = this.conversation[text];
+      break;
+    }
```
I think this would be cleaner if you just initialize `progress` (currently the first property listed under dialog) with the key of wherever the convo starts. Then you would only have to do `this.activeText = this.conversation[this.progress]`

While JS normally does order the keys you're not really supposed to rely on that - different engines do it differently

kauffecup commented on src/Dialogue.js:
```diff
@@ -26,42 +35,77 @@ export default class Dialogue {
+    for (var text in this.conversation) {
+      this.activeText = text;
+      break;
+    }
```
or maybe we also add a property called like `startKey` or something. then for here you would only need to do `this.activeText = this.conversation[this.startKey]`

we would need `startKey` and `progress` to be separate variables - `progress` is what's saved so that exiting/entering an area puts you in the same spot, and killing/restarting the game puts you in the same spot (which I think might be broken right now)

jordantoth replied:
Ah yeah I did break that. Progress will have to be changed to the key of the current block.

kauffecup commented on src/Dialogue.js:
```diff
@@ -26,42 +35,77 @@ export default class Dialogue {
       return this._handleBattle(input, respond, player);
     }
-    return createRegex(this.conversation[this.progress].aliases, false).test(input);
+    return true;//createRegex(this.activeText.aliases, false).test(input);
```
this probably breaks something ;)

jordantoth replied:
This might be what is what's causing items to drop early (if they even are, I don't remember when they were supposed to.
I changed things so the execute function just returns true if execution of the given text block is finished (which at this point pretty much just means any battle is over). I moved the regex creation/testing to a new function.

kauffecup commented on src/Dialogue.js:
```diff
@@ -26,42 +35,77 @@ export default class Dialogue {
+    console.log("no match for input");
+    console.log(input);
+    return false;
+  }
```
before we mergey mergey - we'll remove those console.logs. alsoooo i set up the linting rules to hate "var" - the reason being that vars don't have block scope, they have functional scope - making them behave differently than you'd expect coming from other programming languages

jordantoth replied:
Yeah thought I got all of those. And ya got me, I forgot to run the lint test. I'll check that and it should catch all the vars I used

kauffecup replied:
depending on what editor you're using they're might be a plugin for eslint - i use atom and every time i hit save it runs the lint on that file and let's you know in line what's going down

kauffecup commented on src/entities/Area.js:
```diff
@@ -58,33 +58,85 @@ export default class Area extends _EntityWithInventory {
+              else {
+                // Getting here means we tried to goto a text state that doesn't exist in the conversation
+                // AKA mismatched key, check for typos in JSON.
+                // If Jordan ever makes a JSON form for Patrick this should cease to be an issue
```
yeah it would be dope if we made some kind of "Creator" thing that poops out the JSON files

kauffecup commented:
I'm confused about `area_02`'s conversation still being an array - is that what you meant by "Text files still need to be updated" ?

The only other thing is that i think saving/reloading and exiting an area and coming back to the same spot might be broken. My suggestion in an earlier comment was to keep the current key in the `this.progress` variable. The other big positive side effect of that is you wouldn't need to replace every use of `this.conversation[this.progress]` with `this.activeText`. You would only need to maintain the `this.progress` variable when calling `progressConversation` and those other things would follow

We'd also need to keep track of the starting point - the initial key (instead of relying on the order in the JSON file) - in its own variable somewhere so that `restart` can work

jordantoth commented:
"I'm confused about area_02's conversation still being an array - is that what you meant by "Text files still need to be updated" ?" Yes. I just got it working with area_01

kauffecup commented:
![bitmoji](https://render.bitstrips.com/v2/cpanel/9686055-128732286_4-s4-v1.png?palette=1&width=246)

jordantoth commented:
That

That is horrifying

kauffecup committed:
store current dialogue key in `this.progress` (which is already saved). this allows going and coming back to areas to function and allows killing the game and restarting to work. this changes all uses of `this.activeText` back to `this.conversation[this.progress]`. also restructured the interaction with area a little bit. dialogue.execute() now returns the string of the next key or false (hacky, i know), but this simplifies the logic in Area's execute loop (more similar to what it was before)

kauffecup committed:
fix linting errors

kauffecup committed:
convert area 2 to new fancy format, add support for battles to specify a progression (for now just one - not a map), and add support for dialogues to specify if they can complete the conversation as a part of their map.

kauffecup committed:
fix unit tests. only one needed to be updated - the rest were fixed by updating the JSON files which is pretty sweet

kauffecup commented:
Changes I just made:
- Store the current conversation key in the `this.progress` variable. As a side effect of this, I removed all references to `this.activeText` and replaced them with `this.conversation[this.progress]`
- Added a `resetKey` to simplify the logic for resetting an area (only happens if everyone dies)
- I made `dialogue.execute(input)` either return the key of the next area if there's a match (the return of `getNextTextKey`) or false. `dialogue.progressConversation` now takes the `nextTextKey` and the internals of the method are identical to `goToTextState`. As a side effect of this, the logic in area's execute loop is a little simpler
- The conversation objects can specify a `complete` key in their progression map. The new logic for whether or not the dialogue complete is "if there's no progression map or the current key is 'complete'"
- If the conversation object contains a battle it specifies progression but just as a string instead of an object. The implication of this is that winning a battle plops you into a new spot. I think this is rather kosher but lemme know if you disagree (it makes the code in Dialgue.execute a little hacky)

I also fixed the linting errors and got unit tests to pass

kauffecup committed:
fix tiny bug in requiresitem caused by changes i just pushed
